### ISSUE TITLE
df.apply does not return column names as strings

### ISSUE THREAD

**Issue by carlosssanchez** [modin-project/modin#469]
### System information
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Manjaro Linux
- **Modin installed from (source or binary)**: pip
- **Modin version**: 0.3.1
- **Python version**: 3.7.2
- **Exact command to reproduce**:
```
df = pd.DataFrame( [ [1, 5], [2,4],[3,3],[4,2],[5,1]], columns=['normal','reversed'] )
df['normal'].name
OUT:  
normal                                                                                                                                              
# Printing causes it to output None below 
df.apply( lambda col: print( col.name ) ) 
OUT:
normal      None
reversed    None
dtype: object

#Removing print gives different results
df.apply( lambda col: type(col.name) ) 
OUT:
normal      <class 'numpy.int64'>
reversed    <class 'numpy.int64'>

df.apply( lambda col: str(col.name)   )
normal      0
reversed    1
dtype: object

df.apply( lambda col: col.name[0] ) # Error copied below
```

### Describe the problem
The object returned by col.name is expected to be a string, but instead is returned as a strange numpy int object. This may be related to issue #432. 

### Source code / logs

 File "/home/c/.local/lib/python3.7/site-packages/modin/pandas/dataframe.py", line 781, in apply
    query_compiler = self._query_compiler.apply(func, axis, *args, **kwds)
  File "/home/c/.local/lib/python3.7/site-packages/modin/data_management/query_compiler/pandas_query_compiler.py", line 2397, in apply
    return self._callable_func(func, axis, *args, **kwargs)
  File "/home/c/.local/lib/python3.7/site-packages/modin/data_management/query_compiler/pandas_query_compiler.py", line 2534, in _callable_func
    return self._post_process_apply(result_data, axis)
  File "/home/c/.local/lib/python3.7/site-packages/modin/data_management/query_compiler/pandas_query_compiler.py", line 2417, in _post_process_apply
    internal_index = self.compute_index(0, result_data, True)
  File "/home/c/.local/lib/python3.7/site-packages/modin/data_management/query_compiler/pandas_query_compiler.py", line 138, in compute_index
    old_blocks=old_blocks,
  File "/home/c/.local/lib/python3.7/site-packages/modin/engines/base/block_partitions.py", line 531, in get_indices
    if len(self._partitions_cache.T)
  File "/home/c/.local/lib/python3.7/site-packages/modin/engines/base/block_partitions.py", line 530, in <listcomp>
    [idx.apply(func).get() for idx in self._partitions_cache.T[0]]
  File "/home/c/.local/lib/python3.7/site-packages/modin/engines/ray/pandas_on_ray/remote_partition.py", line 28, in get
    return ray.get(self.oid)
  File "/home/c/.local/lib/python3.7/site-packages/ray/worker.py", line 2132, in get
    raise value
ray.worker.RayTaskError: ray_worker:modin.engines.ray.pandas_on_ray.remote_partition.deploy_ray_func() (pid=30510, host=cliff-pc)
ray.worker.RayTaskError: ray_worker:modin.engines.ray.pandas_on_ray.axis_partition.deploy_ray_func() (pid=30512, host=cliff-pc)
  File "/home/c/.local/lib/python3.7/site-packages/modin/engines/ray/pandas_on_ray/axis_partition.py", line 84, in deploy_ray_func
    return func(*args)
  File "/home/c/.local/lib/python3.7/site-packages/modin/engines/base/axis_partition.py", line 187, in deploy_axis_func
    result = func(dataframe, **kwargs)
  File "/home/c/.local/lib/python3.7/site-packages/modin/data_management/query_compiler/pandas_query_compiler.py", line 163, in helper
    return pandas_func(df, **kwargs)
  File "/home/c/.local/lib/python3.7/site-packages/modin/data_management/query_compiler/pandas_query_compiler.py", line 2531, in <lambda>
    lambda df: callable_apply_builder(df, func, axis, index, *args, **kwargs)
  File "/home/c/.local/lib/python3.7/site-packages/modin/data_management/query_compiler/pandas_query_compiler.py", line 2526, in callable_apply_builder
    result = df.apply(func, axis=axis, *args, **kwargs)
  File "/home/c/.local/lib/python3.7/site-packages/pandas/core/frame.py", line 6014, in apply
    return op.get_result()
  File "/home/c/.local/lib/python3.7/site-packages/pandas/core/apply.py", line 318, in get_result
    return super(FrameRowApply, self).get_result()
  File "/home/c/.local/lib/python3.7/site-packages/pandas/core/apply.py", line 142, in get_result
    return self.apply_standard()
  File "/home/c/.local/lib/python3.7/site-packages/pandas/core/apply.py", line 248, in apply_standard
    self.apply_series_generator()
  File "/home/c/.local/lib/python3.7/site-packages/pandas/core/apply.py", line 277, in apply_series_generator
    results[i] = self.f(v)
File "<ipython-input-7-04dbc5d9f0bf>", line 1, in <lambda>
TypeError: ("'int' object is not subscriptable", 'occurred at index 0')

**Comment by devin-petersohn** [comment]
@carlosssanchez, thanks for posting this!

You are correct that this is related to #432, and the fix will likely be something similar.

This is a consequence of the separation of indexing structures. In this case, we are not properly communicating the column names into the partitions. We keep these structures separate to avoid copying multiple times. The internal indices are kept as `RangeIndex` objects because of their fixed size, so that is why it is returning a `numpy.int64` object as the `type`. Because we have not converted them to the correct indices, it is failing and presenting itself this way.

Thank you for posting this, we are always working on making Modin more robust, especially with `apply`.

**Comment by devin-petersohn** [comment]
After a big rewrite and restructuring of the index objects, this was resolved in #721. Feel free to reopen if the problem persists.
